README.txt
README.TXT -  NPOESS Preparatory Project (NPP) Tropical Cyclone (TC) Science Algorithm (SA) README.TXT file
             
              Author: Scott Longmore, Cooperative Institute for Research in the Atmosphere (CIRA) 
              Contact: Scott.Longmore@colostate.edu
              Last Edit: 2017-09-26


Version Number: v2.0.2

Release Date: alpha 

Configuration Target:

  - Source Directory Structure

    src/
       makefile - Master makefile
       gfspack/ - gfspack code
       ShortTermTrack/ - ShortTermTrack code
       afdeck/ - afdeck code
       satcenter/ - satcenter code
       oparet/ - oparet TC wind estimation code
       common/ - common source code 

    scripts/
       bash/
           NPP_TC.sh - Master control script
           NPP_TC_LIB.sh - Library functions
           NPP_TC_GRIB2PACK.sh - Converts GFS grib2 files to binary, then pack utilizing bin2pack.x
           NPP_TC_MKCONFIG.sh - Create query configuration file for MIRS query data script
                             files to create the NPP_TC.PCF
           NPP_TC_CRON.sh - crontab script that sets time, rundir and executes NPP_TC.sh
    
       python/ - python scripts
          mirs_tc/ - MIRS scheduling software
             mirs_tc.py - scheduling script for MIRS, links latest ATCF, GFS, and 
                          MIRS/NPP data to current run directory, runs NPP_TC.sh, 
                          eventually will replace NPP_TC.sh
          pool/ - MIRS data subset extraction code
             data_pool.py - Extracts a subset of MIRS data from netCDF files to the query subdirectory
          convert/ - converts XYA/RZA files to netCDF
            convert2netCDF.py - netCDF converter
          plot/ - TC wind estimation plotting software
            main_read_plot_xya.py - Plots Wind Field Estimation File
          common/ - common python source code

    bin/ - compiled binaries
       bin2pack.x - converts GFS GRIB to pack format
       ShortTermTrack_mirs.x - find track from current or previous synoptic time
       afdeck.x - determines coordinates from track file
       satcenter.x -  determine satelite scanline time closest to storm
       oparet.x - wind intensity estimation algorithm 
   
    etc/ - configuration and supplemenatry data files
          MIRS_TC.json - master MIRS-TC configuration file 
          MIRS_TC_POOL.ini - configuration file for data pool
          oparet.cfg - scale, offset, and valid values for oparet input data
          plots_config.txt - ploting configuration file
          exitCodes.txt - general and script/executable specific error codes
          MIRS_TC_RT.json - configuration file for task scheduler
          MIRS_TC.crontab - scheduling script crontab
          MIR_TC_CRON.sh - scheduling crontab script
    
    doc/ - documentation files

    test/ - component test directory

  - Working Directory Structure

    <WORKDIR>/ - Working Directory

       MIRS_TC.PCF - Process Control File, including all path, directory, external program, variables, settings, data filename, etc
       MIRS_TC.PSF - Process Status File
       MIRS_TC.LOG - Process Log File

       data/ - Input data files: storm track (deck text), GFS (grib), and MIRS ATMS (netCDF) 

          <aBaSnYYYY.dat> - Input adeck file 
          <gfs analysis file> - GFS analysis file (grib or pack)
          <mirs data files> - MIRS data files (netCDF)

       model/ - Conversion of GFS grib to pack files
          *.bin - variable bin files
          AVN.DAT - GFS pack file
          
       database/ - MIRS subset data

       <BaSnYYYY>/ - Storm processing sub-directories
                     Ba - Basins: [al|ep|wp|cp|io|sh] 
                     Sn - Storm Number
                     YYYYY - Year

           ShortTermTrack_mirs.x - Find track from current or previous synoptic time
              <aBaSnYYYY.dat> - Input adeck file 
              <BaSnYYYY.inp>  - Output storm track file
             
           afdeck.x - Determines coordinates from track file
              <BaSnYYYY.inp> - Input storm track file
              COORDINATES - Output coordinates file

           satcenter.x - Determine satelite scanline time closest to storm
              <BaSnYYYY_YYYYMMDD>.<variable> - Input variable files from pool query (renamed with prefix)
              TIMES - Output time file

           oparet.x - Determines wind fields from MIRS satelite data and GFS model boundary conditions
              AVN.DAT - Input GFS model data file
              <variable.txt> - Input variable files linked to <BaSnYYYY_YYYYMMDD>.<variable> 
              COORTIMES - Input COORDINATE/TIMES concatenated file
              <BaSnYYYY_YYYYMMDD>.LOC - Output Location file 
              <BaSnYYYY_YYYYMMDD>.XYA - Output Wind Field Estimation file 
              <BaSnYYYY_YYYYMMDD>.STA - Output Statistics file 
              <BaSnYYYY_YYYYMMDD>.RZA - Output Variable file 
              <BaSnYYYY_YYYYMMDD>.FIX - Output Intensity/Size Estimation
              <BaSnYYYY_YYYYMMDD>.AFX - Output Aid File  
              <BaSnYYYY_YYYYMMDD>.log - Output Log file 

            main_read_plot_xya.py - Plots Wind Field Estimation File
              <BaSnYYYY_YYYYMMDD>.XYA - Input Wind Field Estimation file 
              <BaSnYYYY_YYYYMMDD>_T_ATMS_250mb.png -    Output 250mb temperature image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_200mb.png - Output 200mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_250mb.png - Output 250mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_300mb.png - Output 300mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_400mb.png - Output 400mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_500mb.png - Output 500mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_600mb.png - Output 600mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_700mb.png - Output 700mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_800mb.png - Output 800mb wind field image file

       output/
         TC-<BaSnYYYY>_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.AFX - AMSU TC AFX fix file
         TC-<BaSnYYYY>_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS_<SnB>_FIX - AMSU TC (JTWC) fix file
         TC-<BaSnYYYY>-XYA_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.nc - XYA netCDF files 
         TC-<BaSnYYYY>-RZA_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.nc - RZA netCDF files 
         TC-<BaSnYYYY>-<field>-<level>_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.png - Storm image files 

       log/
         NPP_TC.LOG - Process Log File 
         <subprocess>.log - Sub-process log files

Reference Documents:
  


Compilation Instructions:

  The master makefile is located in the src/ directory. To compile simply run

     make

  To install executables in the bin/ directory

    make install

  - Compiler Warnings

    No compiler warning with gcc/gfortran 4.4.7-1

Product Control File Description:

    The following data files are needed for NPP TC wind estimation algorithm:

       -  The latest adeck storm track text files avaiable from NHC/JTWC 
       -  The latest GFS model analysis file in grib format 
       -  The latest set of MIRS netCDF files within the last 16hr

Product Status File Description:
      
    The following products will be generated and staged in the working
    directory output/ subdirectory

       TC-<BaSnYYYY>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.<EXT> - Product files for each active storm defined in PSF
       TC-<BaSnYYYY>-<field>-<level>__sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.png - Product image for each active storm files defined in PSF

Production Rule Definitions:

    To be added


RELEASE_NOTES.TXT
RELEASE_NOTES.TXT -  NPOESS Preparatory Project (NPP) Tropical Cyclone (TC) Science Algorithm (SA) RELEASE_NOTES.TXT file
             
Author: Scott Longmore, Cooperative Institute for Research in the Atmosphere (CIRA) 
Contact: Scott.Longmore@colostate.edu
Last Edit: 2017-09-26

2017/09/26  v2.0.2 - Fix plots coastline/boundries/meridians/parallels not being show, increased map resolution to high
2017/09/25  v2.0.1 - Fix for query file gather for correct ordered files, scaling/offset/missing values corrections 
2017/07/10  v2.0.0 - New python driver (mirs_tc.py), new data pool (scripts/python/pool), new operat (src/operat) 
2015/05/15  v1.5.4 - Removed legacy version of nbevs routine insrc/oparet/nbevs.f90, updated besubs.f90 
2015/05/15  v1.5.3 - OPERATIONAL RELEASE
                   - Logging update requested by NDE, WARNING issued when pool_query.py finds no matching records,d
                     sub-programs that exit normally but do not produce output produce WARNING. 
                   - QC.TXT link issues for multiple storms fixed
                   - oparet coefficients updated, minor updates
                   - JPSS satellite variables (for future) commented out
2015/02/02  v1.5.2 - GFS file naming update 
2014/12/10  v1.5.1 - oparet fix for SH storms
                     ShortTermTrack fix for invest storms (missing positions)
                     netCDF global attribute fixes/updates 
2014/11/09  v1.5.0 - Added ATMS QC > 1 for +20% of retrievals, exits algorithm 
                   - Fix for M12 of ShortTermTrack to current synoptic time
                   - ATMS time windows adjusted, gettimes.sh replaced with datetimes.py
		   - XYA/RZA netCDF CF compliance fixes, standard names, valid range, units
		   - JTWC ATCF/FIX filename convention added to products
                   - npp_tc.py scheduler added, eventually will replace master script NPP_TC.sh
2014/07/30  v1.4.2 - Problems with new graphics code on newer ATMS MIRS data 
                     reverted to previous version of plotting code
2014/07/20  v1.4.1 - Added non-zero exit code status in ShortTermTrack for corrupt ATCF files 
                     Updated plotting code, graphics fill more of the frame
2014/07/03  v1.4.0 - Algorithm replaced MIRS ATMS cloud liquid water (clw) with liquid water path (lwp) 
                     to incorporate precipitating (eyewall/rainband) retrievals and added bias correction.
                   - Some sub-program (ShortTermTrack, Oparet) log designators changed from WARNING to INFO 
                     for non-critical exceptions (00 or -12 ATCF fix missing, swath too far from storm, etc) 
2014/05/01  v1.3.4 - Removed dependency on MIRS ATMS attributes with data_pool and oparat changes
                     Added etc/oparet.cfg for correct scaling and offsets
                     Removed TC Sandy case setup/clean scripts, TC test case scripts included with 
                     seperate case data tar file
2014/03/06  v1.3.3 - PCF validator, removes older adeck/gfs files, fix for leap year in NPP_TC_GETTIMES.sh
2014/02/08  v1.3.2 - netCDF bug fixes, minor updates 
                   - ERROR to WARNING messages for sub-programs inside storm loop, so if one TC fails,
                     remaining TC's are processed 
2014/02/02  v1.3.1 - netCDF converter for RZA files
2014/01/27  v1.3.0 - netCDF converter for XYA files, MiRS file fix
2014/01/07: v1.2.3 - ATCF (adeck), GFS grib file naming, product version in file names
2013/11/14: v1.2.2 - Data test case inclusion
2013/11/08: v1.2.1 - PCF formatting, required NDE and data input variable upgrades 
                   - Data, temporary sub-directory hierarchy handling
                   - Static variables to master script, external and data variables to PCF
                   - Date format/handling between NDE and master script
                   - Executable/script location execution handling from working to operation directory 
                   - PSF output file location and naming is now absolute 
                   - LOG file location to top working directory
2013/10/31: v1.2.0 - NDE framework upgrades: PCF, PSF, logging, exception handling
                   - GFS grib to pack file converter
2013/06/25: v1.1.0 - NDE initial version
2013/06/14: v1.0.0 - Initial version


src/oparet/oparet/io.f90
      module io 
        implicit none

        ! Variables for input file names
        character(len=25) :: fntinp = 'COORTIMES'
        character(len=25) :: fnninp = 'AVN.DAT'
        character(len=25) :: fncinp = 'oparet.cfg'
        character(len=25) :: fnparam = 'oparet_params.cfg'

        ! Variables for input filenames read from config file
        character(len=25) :: fl_coeff_cvmx
        character(len=25) :: fl_coeff_cpmn
        character(len=25) :: fl_coeff_cr34
        character(len=25) :: fl_coeff_cr50
        character(len=25) :: fl_coeff_cr64
        character(len=25) :: instr
        ! number of coeffs
        integer :: ncoefs  !number of coefficients for
                                           !each var 

        ! Variables for output file names
        character(len=25) :: fnlog = 'oparet.log'
        character(len=25) :: fnloc
        character(len=25) :: fnnpk,fnapk
        character(len=25) :: fnrza,fnxya
        character(len=25) :: fnsta,fnfix,fnafx

        ! Variable for temporary file names
        character(len=25) :: fntemp

        ! Variables for output file headers
        character(len=90) :: coord,times

        ! Input file numbers
        integer,parameter :: lutinp = 21 
        integer,parameter :: luninp = 22 
        integer,parameter :: luainp = 23
        integer,parameter :: luparam = 24
        integer,parameter :: lucoef = 25

        ! Output file numbers
        integer,parameter :: luloc = 31 
        integer,parameter :: lulog = 32
        integer,parameter :: lunpk = 33
        integer,parameter :: luapk = 34
        integer,parameter :: lurza = 35
        integer,parameter :: lusta = 36
        integer,parameter :: lufix = 37
        integer,parameter :: luafx = 38
        integer,parameter :: luxya = 39
 
        ! Input Attributes 
        type inpAtt
          real :: latScale,latOffset,latValid
          real :: lonScale,lonOffset,lonValid
          real :: prsScale,prsOffset,prsValid
          real :: clwScale,clwOffset,clwValid
          real :: tpwScale,tpwOffset,tpwValid
          real :: tmpScale,tmpOffset,tmpValid
          real :: vprScale,vprOffset,vprValid
        end type inpAtt 

        !parameters reaqd from config file
        real    :: scale_rv
        real    :: qcFailValue
        logical :: use_tv = .FALSE. 
        

        type (inpAtt) :: inpAtts 

        save

        contains

        subroutine read_params_config
          integer :: open_status
          integer :: io_status
          integer :: ierr

          integer :: use_tv_int, i, nlines
          character(len=80) :: iline

          ! count number of lines in config file
          open(unit=luparam,file=fnparam,status='OLD',iostat=open_status)
          nlines = 0
          do i = 1, 10000000
            read (luparam,103, end = 900) iline
            nlines = nlines+1
          enddo
  900     rewind(luparam)

          ! read params from config file
          do i = 1, nlines
            read (luparam, 103) iline

              if (iline(1:6) .eq. 'use_tv') then
                read(iline,102) use_tv_int
                if (use_tv_int == 1) then
                    use_tv = .TRUE.
                endif
              endif
              
             if (iline(1:6) .eq. 'ncoefs') then
                read(iline,102) ncoefs
              endif
              
              if (iline(1:8) .eq. 'scale_rv') then
                read(iline,101) scale_rv
              endif
              
              if (iline(1:11) .eq. 'qcFailValue') then
                read(iline,101) qcFailValue
              endif
              
              if (iline(1:8) .eq. 'vmx_coef') then
                fl_coeff_cvmx=adjustl(trim(iline(26:)))
              endif
        
              if (iline(1:8) .eq. 'pmn_coef') then
                fl_coeff_cpmn=adjustl(trim(iline(26:)))
              endif
              
              if (iline(1:8) .eq. 'r34_coef') then
                fl_coeff_cr34=adjustl(trim(iline(26:)))
              endif
              
              if (iline(1:8) .eq. 'r50_coef') then
                fl_coeff_cr50=adjustl(trim(iline(26:)))
              endif
              
              if (iline(1:8) .eq. 'r64_coef') then
                fl_coeff_cr64=adjustl(trim(iline(26:)))
              endif
              
              if (iline(1:5) .eq. 'instr') then
                instr=adjustl(trim(iline(26:)))
              endif


          enddo

  101     format(26x,f18.5)
  102     format(26x,i20)
  103     format(a80)
  104     format(26x,a80)

          close(luparam)
       
          print *, "config params :"
          print *, "nlines in config file ", nlines 
          print *, "use_tv  = " ,       use_tv 
          print *, "scale_rv  = ",      scale_rv
          print *, "qcFailValue  = ",   qcFailValue 
          print *, "ncoefs  = ",   ncoefs
          print *, "fl_coeff_cvmx  = ",   fl_coeff_cvmx
          print *, "fl_coeff_cpmn  = ",   fl_coeff_cpmn
          print *, "fl_coeff_cr34  = ",   fl_coeff_cr34
          print *, "fl_coeff_cr50  = ",   fl_coeff_cr50
          print *, "fl_coeff_cr64  = ",   fl_coeff_cr64
          print *, "instr  = ",   instr

        end subroutine


        subroutine read_config 
          integer :: open_status
          integer :: io_status
          integer :: ierr

          open(unit=luainp,file=fncinp,status='OLD',iostat=open_status)
          if(open_status /= 0) then 
            write(*,*) 'WARNING: could not open input attribute file'
            ierr=66
            call exit(ierr)
          endif 

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%latScale,inpAtts%latOffset,inpAtts%latValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%latScale,inpAtts%latOffset,inpAtts%latValid

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%lonScale,inpAtts%lonOffset,inpAtts%lonValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%lonScale,inpAtts%lonOffset,inpAtts%lonValid

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%prsScale,inpAtts%prsOffset,inpAtts%prsValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%prsScale,inpAtts%prsOffset,inpAtts%prsValid

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%clwScale,inpAtts%clwOffset,inpAtts%clwValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%clwScale,inpAtts%clwOffset,inpAtts%clwValid

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%tpwScale,inpAtts%tpwOffset,inpAtts%tpwValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%tpwScale,inpAtts%tpwOffset,inpAtts%tpwValid

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%tmpScale,inpAtts%tmpOffset,inpAtts%tmpValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%tmpScale,inpAtts%tmpOffset,inpAtts%tmpValid

          read(luainp,'(11X,F15.5,1X,F15.5,1X,F15.5)',iostat=io_status)&
          inpAtts%vprScale,inpAtts%vprOffset,inpAtts%vprValid
          ! write(*,'(3F15.5)') &
          ! inpAtts%vprScale,inpAtts%vprOffset,inpAtts%vprValid

          close(luainp)

        end subroutine read_config 

        subroutine init_output
          use params
          use sinfo1
          use sinfo2

          character(len=3) :: lab3
          integer :: idayr,iprt,ios
          
          ! ++ Name for loc file (AMSU footprint locations)
          fnloc        = 'LX0000_000000.DAT'
          fnloc( 2: 2) = atcfid(1:1)
          fnloc( 3: 6) = atcfid(3:6)
          fnloc(18:19) = '99'
          write(fnloc(8:13),226) imon,iday,itime
          226 format(3(i2.2))

          if (igenfn .eq. 1) then
            fnloc='  '
            fnloc(1:6) = csatid
            fnloc(7:10) = '.LOC'
          endif

          ! ++ Name for rza file (2-D AMSU retrieval output file)
          fnrza        = 'RX0000_000000.DAT'
          fnrza( 2: 2) = atcfid(1:1)
          fnrza( 3: 6) = atcfid(3:6)
          fnrza(18:19) = '99'
          write(fnrza(8:13),226) imon,iday,itime

          if (igenfn .eq. 1) then
            fnrza='  '
            fnrza(1:6) = csatid
            fnrza(7:10) = '.RZA'
          endif

          ! ++ Name for xya file (3-D AMSU retrieval output file)
          fnxya        = 'AX0000_000000.DAT'
          fnxya( 2: 2) = atcfid(1:1)
          fnxya( 3: 6) = atcfid(3:6)
          fnxya(18:19) = '99'
          write(fnxya(8:13),226) imon,iday,itime

          if (igenfn .eq. 1) then
            fnxya='  '
            fnxya(1:6) = csatid
            fnxya(7:10) = '.XYA'
          endif

          ! ++ Name for sta file (TC statistics file)
          fnsta        = 'SX0000_000000.DAT'
          fnsta( 2: 2) = atcfid(1:1)
          fnsta( 3: 6) = atcfid(3:6)
          fnsta(18:19) = '99'
          write(fnsta(8:13),226) imon,iday,itime

          if (igenfn .eq. 1) then
            fnsta='  '
            fnsta(1:6) = csatid
            fnsta(7:10) = '.STA'
          endif

          ! ++ Create file name for fix file
          fnfix        = 'FX0000_000000.DAT'
          fnfix( 2: 2) = atcfid(1:1)
          fnfix( 3: 6) = atcfid(3:6)
          fnfix(18:19) = '99'
          write(fnfix(8:13),226) imon,iday,itime

          if (igenfn .eq. 1) then
            fnfix='  '
            fnfix(1:6) = csatid
            fnfix(7:10) = '.FIX'
          endif

          ! ++ Create file name for afx file
          fnafx        = 'GX0000_000000.DAT'
          fnafx( 2: 2) = atcfid(1:1)
          fnafx( 3: 6) = atcfid(3:6)
          fnafx(18:19) = '99'

          write(fnafx(8:13),226) imon,iday,itime
          if (igenfn .eq. 1) then
            fnafx='  '
            fnafx(1:6) = csatid
            fnafx(7:10) = '.AFX'
          endif

          ! ++ Name for npk file (packed NCEP output file)
          if (itime .eq. 6 .or. itime .eq. 18) then
            rawdid = 'Y'
          else
            rawdid = 'X'
          endif

          if (itime .ge. 12) then
            idayr = iday+50
          else
            idayr = iday
          endif

          fnnpk       = ' '
          fnnpk(1:1)  = 'N'
          fnnpk(2:2)  = atcfid(1:1)
          fnnpk(3:6)  = atcfid(3:6)
          fnnpk(7:21) = '_X0000_PACK.DAT'
          fnnpk(22:23)= '99'
          write(fnnpk(8:12),300) rawdid,imon,idayr
          300 format(a1,i2.2,i2.2)

          if (igenfn .eq. 1) then
            fnnpk='  '
            fnnpk(1:6) = csatid
            fnnpk(7:10) = '.NPK'
          endif

          ! ++ Name for apk file (packed 3-D AMSU retrieval output file)
          ! on lat,lon,P grid
          fnapk=fnnpk
          fnapk(1:1) = 'A'

          if (igenfn .eq. 1) then
            fnapk='  '
            fnapk(1:6) = csatid
            fnapk(7:10) = '.APK'
          endif

          ! ++ Write file names to log file
          lab3   =  'loc'
          fntemp = fnloc
          iprt   = iploc
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'rza'
          fntemp = fnrza
          iprt   = iprza
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'xya'
          fntemp = fnxya
          iprt   = ipxya
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'sta'
          fntemp = fnsta
          iprt   = ipsta
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'fix'
          fntemp = fnfix
          iprt   = ipfix
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'afx'
          fntemp = fnafx
          iprt   = ipfix
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'apk'
          fntemp = fnapk
          iprt   = ipapk
          write(lulog,218) lab3,fntemp,iprt

          lab3   =  'npk'
          fntemp = fnnpk
          iprt   = ipnpk
          write(lulog,218) lab3,fntemp,iprt

          218 format(a3,' data output file: ',a25,' write flag=',i1)

          ! ++ Open the output files, if necessary 
          ! Note: Fix file is opened later if needed, depending
          ! on the outcome of the TC retrieval

          if (iploc .eq. 1) then
            fntemp=fnloc
            open(unit=luloc,file=fnloc,form='formatted',status='unknown',iostat=ios)
            if(ios.ne.0) call open_error
          endif

          if (iprza .eq. 1 .and. iaxsym .eq. 1) then
            fntemp = fnrza
            open(file=fnrza,unit=lurza,form='formatted',status='unknown',iostat=ios)
            if(ios.ne.0) call open_error
          endif

          if (ipxya .eq. 1) then
            fntemp=fnxya
            open(unit=luxya,file=fnxya,form='formatted',status='unknown',iostat=ios)
            if(ios.ne.0) call open_error
          endif

          if (ipsta .eq. 1 .and. iaxsym .eq. 1) then
            fntemp = fnsta
            open(file=fnsta,unit=lusta,form='formatted',status='unknown',iostat=ios)
            if(ios.ne.0) call open_error
          endif

          if (ipapk .eq. 1) then
            fntemp=fnapk
            open(unit=luapk,file=fnapk,form='formatted',status='unknown',iostat=ios)
            if(ios.ne.0) call open_error
          endif

          if (ipnpk .eq. 1) then
            fntemp=fnnpk
            open(unit=lunpk,file=fnnpk,form='formatted',status='unknown',iostat=ios)
            if(ios.ne.0) call open_error
          endif

        end subroutine init_output 

        ! Error processing
        subroutine open_error 
          integer :: ierr

          write(lulog,990) fntemp
          990 format(/,' Error opening file ',a25)
          write(*,995) fntemp
          995 format('WARNING: could not open file ',a25)
          ierr=66
          call exit(ierr)

        end subroutine open_error

      end module io 
